### ISSUE TITLE
Custom character as separator for multiple Artists

### ISSUE THREAD

**Issue by nebula-it** [deluan/navidrome#238]
As discussed [here](https://github.com/deluan/navidrome/issues/203#issuecomment-619618356_) it would be a good idea to implement a setting where end user can select a custom character that will be used to split multiple artist into separate values.

### Current Behavior 
A song has Contributing Artist value `Artist1, Artist2`. Navidrome reads the artist as a single value with name `"Artist1, Artist2"`

### Expected Behavior 
A song has Contributing Artist value `Artist1, Artist2`. Navidrome reads the artist and recognizes there are two or multiple artists separated by comma and song show up under the Artist page of `Artist1` and `Artist2`. 
The separator character can be selected (comma in the above example) by end user depending on how their library is tagged.

**Comment by deluan** [comment]
Thanks, I will try to work on this after #211

**Comment by ElleshaHackett** [comment]
I would like to see this available for multiple custom seperators, not limited to one or several single characters. For uses such as " feat. ", " featuring ", " vs ", " vs." and " versus " and possible " & ", " and ". All should be case insensitive (or optional) and after separation the values should be trimmed.

Maybe we can allow regular expressions for separation? Keep in mind that introduces some possible security implications like [ReDoS](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS).

**Comment by deluan** [comment]
Yes, the idea is to allow multiple words as separators. Not sure if Regex is any good here (and have the security implications as you said)

**Comment by certuna** [comment]
Bear in mind that there *is* a standard separator: `/` for id3v2.3, and the null character (0x00) for id3v2.4. Vorbis Comments (used for FLAC) just has multiple ARTIST tags. Other separators like `;,&` are technically nonstandard hacks.

The fancier the separators are, the more chance of false positieves. Some examples that get can get split accidentally:
"Earth, Wind & Fire"
"AC/DC"  - this was cited back in the days as the big reason why id3v2.3 was terrible and the spec change for v2.4 was needed :)
"Ånd"
"Angels Versus Animals"
"Little Feat"

Interesting discussions:
https://hydrogenaud.io/index.php?topic=112634.0
https://emby.media/community/index.php?/topic/5731-question-on-music-tags-and-delimiters/
https://community.mp3tag.de/t/multiple-genres-in-1-genre-tag/17159

**Comment by deluan** [comment]
`ffmpeg` (used to extract the tags) does not support id3v2.4 multi-valued tags: https://trac.ffmpeg.org/ticket/6949. I just tested with the latest version and it basically ignore everything after the `null` separator.... 

To make this work, I'll need to find an alternative way to read tags (at least id3 tags, have to test with vorbis and mp4), that is robust enough and does not add extra dependencies to the project....

**Comment by certuna** [comment]
That is pretty terrible actually - three years and still not fixed?

**Comment by certuna** [comment]
By the looks of it, the ffmpeg devs don't care, so the solution is either find another tag library or write a commit for ffmpeg (how's your C?) - what's the least amount of work?

**Comment by deluan** [comment]
Both options are huge tasks (for my limited time). I tried replacing `ffmpeg` before, but could not find a library that would cover all formats and all required info I need from the files (including `duration` and `bitrate`). I also use https://github.com/dhowden/tag to extract pictures (and it gets null-separated multi-valued fields correctly), but it does not provide all info required.

I'll take a look in the ffmpeg codebase, but I don't anticipate a easy change to fix this...

**Comment by certuna** [comment]
So [https://github.com/taglib/taglib](taglib) is our new saviour?

**Comment by certuna** [comment]
Is there any reason why there still seems to be (reasonably) recent commits but no releases since 2016?

**Comment by certuna** [comment]
By the way, I see that Picard has introduced a new custom tag for "artists" (`TXXX:Artists` in id3, `ARTISTS` in Vorbis) to store multivalued artists, presumably so old/noncompliant players don't choke on multivalued fields in the regular "artist" fields `TPE1` and `ARTIST` . An ugly hack I think, but probably worth including this field when multivalued fields are implemented.

**Comment by orlea** [comment]
Will this feature be added in the future?

**Comment by deluan** [comment]
Yes, definitely. Probably next after some big PRs are merged and I finish Smart playlists

**Comment by krateng** [comment]
In the meantime, while Navidrome can only accept one artist, is there something that could be done about the multiple artist tags in Vorbis comments (flac)? If I have a track with

```
ARTIST=K/DA
ARTIST=Seraphine
```

it only shows one of them - would be nice to simply display that as `K/DA;Seraphine` for the purpose of scrobbling.

**Comment by certuna** [comment]
That's a good idea - I'll have a look how hard it is to implement that. BTW this can only be done when using TagLib as a scanner, since ffmpeg will never read the second value.

**Comment by Baswazz** [comment]
I see LMS is supporting `\\` as a separator. This could be an idea also supported with multiple genres.

`Artist1\\Artist2`

**Comment by irdkwmnsb** [comment]
Any progress on this feature?

**Comment by jmbannon** [comment]
id3v2.4 using `albumartists` should be the way multiple artists is implemented, not separators

**Comment by certuna** [comment]
the custom fields `artists` and `albumartists` is a pretty dirty hack for applications that cannot read multivalued tags correctly, it's probably better to just implement multivalued tags outright.

**Comment by tam481** [comment]
This will also help with multilingual tagging where the artist name can be in the native language and perhaps the "ARTISTSORT" tag can be used for the English version of the name etc

**Comment by Xarishark** [comment]
Streaming services have made the best version of this right now. On Deezer the album/song appears on both artists pages and where the artists info is it displays both artists as separate clickable links to go to their artist page. Navidrome currently does not even SHOW that there is a second artists let alone let you click him etc.

**Comment by certuna** [comment]
The advantage that Spotify/Deezer/Tidal have is that they can control the exact formatting of the tags, which is a luxury we don't have. Also, they likely have implemented M2M relations in their underlying DB, which Navidrome doesn't have.

There are basically three main ways that multiple artists are tagged these days:
- multi-value like the standard prescribes, so in id3v2: `TPE1` = "Alice<null>Bob"
- single-value with a non-standard separator, usually semicolon: `TPE1` = "Alice;Bob"
- using two tags, one single-value and one multi-value: `TPE1` = "Alice feat. Bob" and `TXXX:ARTISTS` = "Alice<null>Bob" (it may seem a bit hacky to use two tags to do the work of one, but this is how MusicBrainz Picard tags by default)

All three methods would have to be supported. I currently have a build running with multiple artist support (all three variants). However, due to the current DB/API limitations, the links in the UI only take you to the first artist. But it's a lot better than nothing.

![Screenshot 2023-06-16 at 14 08 49](https://github.com/navidrome/navidrome/assets/62144283/64cd0ac4-e45e-447a-a7b8-770cc0ba49e8)

**Comment by jmbannon** [comment]
Looks awesome @certuna 
For any beets users, I have a PR up for multi-tag support: https://github.com/beetbox/beets/pull/4743

**Comment by certuna** [comment]
One of the tricky things of the two-tags approach is the edge case `ARTIST` = "Alice" and `ARTISTS` = "Bob", "Carol" - what should the application display?

**Comment by jmbannon** [comment]
I thought ARTIST and ALBUMARTIST typically included the join phrase, i.e. "Alice feature Bob", "Alice, Bob", "Alice ft. Bob", etc. The app should use ARTISTS since it's more uniform. ARTIST should be treated more-so as the display artist

**Comment by jmbannon** [comment]
But then again not everyone will have multi-tags, so maybe use ARTIST as a fall-back

**Comment by certuna** [comment]
Exactly, but this assumes ARTIST has consistency with ARTISTS, which may be true for collections fully managed with autotaggers like Picard/beets, but otherwise not guaranteed, for example when someone manually fixes tags.

**Comment by pjadeslandes** [comment]
> I would like to see this available for multiple custom seperators, not limited to one or several single characters. For uses such as " feat. ", " featuring ", " vs ", " vs." and " versus " and possible " & ", " and ". All should be case insensitive (or optional) and after separation the values should be trimmed.
> 
> Maybe we can allow regular expressions for separation? Keep in mind that introduces some possible security implications like [ReDoS](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS).

I know this is super old, but I'm investigating Navidrome because I have exactly this request.

I'm aware of some of the edge cases that exist, but I always imagine there are ways around them.

For example, "AC/DC" is a continuous string with no spaces, so if using / as a delimiter, put spaces around it.

Also the use of & - to me, that's used to separate different artists, while "and" is not. Eg "Marina and The Diamonds" is one artist, while "Beyoncé & Lady Gaga" is two. And so if I want, I can treat "Prince and The Revolution" as one artist or, better for me, as two by changing the tag to "Prince & The Revolution", not because I need to see The Revolution as an artists, but I'll take that if it means that all of Prince's music is in one place, whether he was with one of his bands or not.

One slight tweak to the request: I'd like to pull artists from the track name using a format like "<track title> (feat <ft artist one>, <ft artist two>, etc)".

Am following the main thread on all the most recent updates. Amazing work by all involved, however far all of this gets. Such a shame commercial players like Plex aren't as dedicated to as high a quality of experience. V happy to pay / donate to something like the Navidrome prject, which is so responsive to such specific user requests.

**Comment by certuna** [comment]
- `&` is an extremely unreliable separator. There are way, way too many single artists with this character in their name - from `Years & Years` to `Earth, Wind & Fire` to `Kool & The Gang` to `Milk & Sugar` to `The Mamas & The Papas`. Discogs has over 200,000 (!) artists with `&` in the name, and these are not all duos.
- `/` has less issues, but still has dozens of exceptions, Discogs has over 3,000 artists
- `;` is a very reliable separator, Discogs has only 12 artists

Splitting text strings reliably is very hard, this is why the ARTIST/ARTISTS method is used by for example Picard: ARTIST = "Alice featuring Bob", ARTISTS = "Alice", "Bob", so applications that can handle multiple artists use what's in the ARTISTS tag without having to do complex splitting operations, and oldschool applications that can only handle one artist use the contents of ARTIST.

**Comment by jmbannon** [comment]
My two cents, if `ARTIST` was possible to reliably split on, there would be no need for `ARTISTS`. There will always be an edge case, which is why self hosted music should embrace the `ARTISTS` + `ALBUMARTISTS` multi-tags. I guarantee Spotify, Deezer, etc use something very similar to `ARTISTS` in the form of a database.

**Comment by jmbannon** [comment]
Supporting multiple artists for folks who don't have multi-tags, to put it bluntly, is not feasible, and trying to do so I think contributes to the problem of self-hosted music's messy standard for supporting multiple artists that we see today.

The software that I use that does support multiple artists (Kodi, Jellyfin, MusicBee) treats `ARTIST`/`ALBUMARTIST` as the display artist containing the join phrases (ft. featuring, with, etc) without trying to be split on, and `ARTISTS`/`ALBUMARTISTS` to use strictly for multiple artists. This is the closest standard I've seen that I think should be adopted abroad and in your approach. There is plenty of available software (MusicBrainz Picard currently the best, beets very soon with https://github.com/beetbox/beets/pull/4743) that can add these multi-tags.

Only using multi-tags for multiple artist support will both simplify your PR and help cement this standard. I think it's okay to leave behind folks who refuse to add multi-tags.

(sorry if this comes off as brash, I just want the world to use multi-tags and enjoy multiple artist support like I do :P )

**Comment by certuna** [comment]
It's not so bad.

- Supporting standard-conforming multi-tags is easy.

- Supporting the ARTIST+ARTISTS hack/workaround is also easy, only thing you need to watch out for is inconsistency between the two tags.

- The difficult part is non-standard separators in a single field, the way I've implemented it now is that it only supports the semicolon by default, and if people insist on using other characters like /, they're able to specify them.

**Comment by pjadeslandes** [comment]
I was only thinking of these delimiters as optional characters, I'm aware something like "&" won't work for everyone but I know it would work for me. I'm all for using tags such as ARTISTS, although I use mp3Tag and I don't see it there so I guess I'd have to move apps, which won't kill me! As ever, I'm impressed by how engaged everyone here is and am very appreciative of the work done by @certuna and all other contributors.

**Comment by jmbannon** [comment]
@pjadeslandes I think it's under extended tags

**Comment by certuna** [comment]
BTW I'm just discovering that Musicbrainz Picard by default writes `ARTIST` (single) + `ARTISTS` (multi), but not like you'd expect `ALBUMARTIST` + `ALBUMARTISTS` - it just drops any additional album artists and writes `ALBUMARTIST` as a single field. But it does write multiple artist IDs to the Release Artist ID field...so as usual with non-standard stuff, sloppy implementations everywhere.

**Comment by jmbannon** [comment]
They have it but you have to enable it manually. See the bottom of https://kodi.wiki/view/Music_tagging

**Comment by certuna** [comment]
Yeah I saw that - I've had that for testing, but it's a bit annoying for widespread adoption that it's not on by default.

**Comment by certuna** [comment]
The Navidrome fork I'm using for testing is now open and available for all to try, if you're interested: https://github.com/certuna/navidrome/tree/multi-artists . Still doing more testing before I submit the PR, but so far I see no showstoppers.

**Comment by jmbannon** [comment]
@certuna plz add id3v2.4 multi support for a separator type

**Comment by raudraido** [comment]
> I see LMS is supporting `\\` as a separator. This could be an idea also supported with multiple genres.
> 
> `Artist1\\Artist2`

My library is separated like this and Emby works very well with it. Also it recognizes "feat." syntax in Artist name.

**Comment by makerdegit** [comment]
> The Navidrome fork I'm using for testing is now open and available for all to try, if you're interested: https://github.com/certuna/navidrome/tree/multi-artists . Still doing more testing before I submit the PR to the main Navidrome repo, but so far I see no showstoppers. Scanning speed appears to be unchanged on my 160k song library.
> 
> Four new server options (default):
> 
> * `OnlyAlbumArtists` (`true`): when clicking on the Artists section in the left-side nav bar, either show a list of only the artists that have an album, or a list of all artists including those with only V/A compilation appearances.
> * `Scanner.MultipleArtists` (`true`): if set to false, ND will only use the first track/album artist, like now
> * `Scanner.ArtistSeparators` (`";"`): self explanatory - list of custom characters to split artists on, exactly like with the existing Scanner.GenreSeparators option.
> * `Scanner.RemixerToArtist` (`true`): adds the Remixer to the list of track artists, like Spotify, Tidal and Deezer do. Of course, for this to work, you need to have the Remixer tag ( in id3) properly tagged.`TPE4`

Hello. Up to now, the problem with many artists has still not been resolved in the official version. I've noticed that you've released a version that can solve this problem. However, I'd like to deploy it using Docker. Could you please release a Docker version?

**Comment by makerdegit** [comment]
> The Navidrome fork I'm using for testing is now open and available for all to try, if you're interested: https://github.com/certuna/navidrome/tree/multi-artists . Still doing more testing before I submit the PR to the main Navidrome repo, but so far I see no showstoppers. Scanning speed appears to be unchanged on my 160k song library.
> 
> Four new server options (default):
> 
> * `OnlyAlbumArtists` (`true`): when clicking on the Artists section in the left-side nav bar, either show a list of only the artists that have an album, or a list of all artists including those with only V/A compilation appearances.
> * `Scanner.MultipleArtists` (`true`): if set to false, ND will only use the first track/album artist, like now
> * `Scanner.ArtistSeparators` (`";"`): self explanatory - list of custom characters to split artists on, exactly like with the existing Scanner.GenreSeparators option.
> * `Scanner.RemixerToArtist` (`true`): adds the Remixer to the list of track artists, like Spotify, Tidal and Deezer do. Of course, for this to work, you need to have the Remixer tag ( in id3) properly tagged.`TPE4`

Hello, the issue with recognizing multiple artists has not been resolved in the official version so far. I noticed that you released a version that can fix this problem, but I would like to deploy it using Docker. Could you please release a Docker version?

**Comment by Jucgshu** [comment]
Deluan is working hard on a massive refactoring that includes this feature. So you'd pretty much have to do nothing but to wait to get it :wink:

**Comment by rwa96** [comment]
Having ; in addition to / would solve the AC/DC problem

**Comment by deluan** [comment]
> Having ; in addition to / would solve the AC/DC problem

Sorry, but I don't follow. How `;` will affect AC/DCD problem?

Also note that the separator is ` / ` (surrounded by spaces), which already fixes the AC/DC issue.

**Comment by rwa96** [comment]
The problem is that / is more commonly used than ; in my experience. I have to admit, I didn't know about the spaces, and you are right, AC/DC is not an issue anymore. Now I only have to make an exception for L I F E / L I N E remixes.

**Comment by afontenot** [comment]
> Yes, this is already implemented in #2709

Part of the discussion on this issue was also support for proper multi-valued tags. Is that included in the PR? If so, does it supersede any join phrase (which could be there by accident for those with properly tagged multi-value files)?

Just for clarity, here's an example of album artists tags on a file, as shown by `metaflac`:

```
    comment[4]: ALBUMARTIST=Danger Mouse & Black Thought
    comment[5]: ALBUMARTISTSORT=Danger Mouse & Black Thought
    comment[6]: ALBUMARTISTS=Danger Mouse
    comment[7]: ALBUMARTISTS=Black Thought
```

**Comment by deluan** [comment]
Splitting will only occur if there is no ARTISTS/ALBUMARTISTS tag AND the tag is single valued.

In your example:
- ALBUMARTIST will be used as DisplayArtist in Subsonic clients.
- ALBUMARTISTSORT will be used for sorting
- ALBUMARTISTS will be imported as two separated artists: "Danger Mouse" and "Black Thought"
Navidrome UI will show The album under both artists, and when you open the album, it will show: "Danger Mouse • Black Thought", each artist name will be clickable separatedly.

Some examples:
![Screenshot 2024-11-08 at 9 54 06 PM](https://github.com/user-attachments/assets/d44b842b-f2ed-4d18-a998-455b88e7a45a)
![Screenshot 2024-11-08 at 4 55 04 PM](https://github.com/user-attachments/assets/757884cb-63b8-4212-96de-d7de10ebb255)

Details of the implementation have been discussed in our Discord: https://discord.com/channels/671335427726114836/1249159191608692788

**Comment by jamie6king** [comment]
Thanks, this works great on the frontend! However, it appears it still sends only the first artist to ListenBrainz?

**Comment by deluan** [comment]
> Thanks, this works great on the frontend! However, it appears it still sends only the first artist to ListenBrainz?

Thanks for reporting. Fixed in ed8f5becd. Let me know if it works for you.

**Comment by jamie6king** [comment]
using the latest image from navidrome:pr-2709 on docker and doesn't seem to work? Using Maloja for scrobbling and it only shows the first artist. Might be something to do with it also not displaying multiple artists when playing (e.g. the bottom bar & OS playback area).

**Comment by deluan** [comment]
@jamie6king, this conversation is getting out of topic. Can you please open a new discussion in https://github.com/navidrome/navidrome/discussions/categories/general for us to troubleshoot this?

Thanks.

**Comment by jamie6king** [comment]
apologies, opened a new issue at 3638.

**Comment by deluan** [comment]
This is complete and released in 0.55.0
